### PR TITLE
feature: Add serialization of reference dataset

### PR DESCRIPTION
**Pull request by svotaw**
## Summary
This is in reference to feature request: https://github.com/microsoft/LightGBM/issues/5426

This PR adds APIs for serializing/deserializing Datasets without their data to a byte array, effectively creating a "schema" or "reference" that can be used to create other Datasets.

## Implementation
The existing code for serializing Datasets to file was refactored to be able to go to any generic `BinaryWriter`, whether memory or file. The verbose serialization code was shared as much as possible, splitting methods into Header vs Data components.

Also, a generic `ByteBuffer` was created so that higher languages (e.g. Java) are removed from managing the byte memory of the serialized buffer.

## Test
New C++ tests were created to test both the serialization/deserialization and the new `ByteBuffer` functionality.

### CODE REVIEW

File: src/io/dataset.cpp [hunk 1]
```diff
@@ -19,8 +19,12 @@
 
 namespace LightGBM {
 
+const float Dataset::kSerializedReferenceVersion = 1.0;
```
Comment (StrikerRUS) [inline src/io/dataset.cpp:22]: Maybe use string type like it is used for saved model version?
https://github.com/microsoft/LightGBM/blob/b0137debe6e9cc92b65ec71b0fe8a56ea213c143/src/boosting/gbdt_model_text.cpp#L19

File: src/io/dataset.cpp [hunk 2]
```diff
@@ -1094,6 +1027,120 @@ void Dataset::SaveBinaryFile(const char* bin_filename) {
...
+  for (int i = 0; i < num_total_features_; ++i) {
+    size_of_header +=
+      VirtualFileWriter::AlignedSize(feature_names_[i].size()) +
+      VirtualFileWriter::AlignedSize(sizeof(int));
+  }
+  // size of forced bins
+  for (int i = 0; i < num_total_features_; ++i) {
+    size_of_header += forced_bin_bounds_[i].size() * sizeof(double) +
+      VirtualFileWriter::AlignedSize(sizeof(int));
+  }
```
Comment (StrikerRUS) [inline src/io/dataset.cpp:1085]: Looks like it can be merged into one `for` loop.

File: tests/cpp_tests/test_serialize.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#include <iostream>
+
+using LightGBM::ByteBuffer;
+using LightGBM::Dataset;
+using LightGBM::Log;
+using LightGBM::TestUtils;
+
+TEST(Serialization, JustWorks) {
+  // Load some test data
+  DatasetHandle datset_handle;
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:22]:
```suggestion
  DatasetHandle dataset_handle;
```

File: tests/cpp_tests/test_serialize.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  const char* params = "max_bin=15";
+  int result = TestUtils::LoadDatasetFromExamples("binary_classification/binary.train", params, &datset_handle);
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:24]:
```suggestion
  int result = TestUtils::LoadDatasetFromExamples("binary_classification/binary.train", params, &dataset_handle);
```

File: tests/cpp_tests/test_serialize.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  EXPECT_EQ(0, result) << "LoadDatasetFromExamples result code: " << result;
+
+  Dataset* dataset;
+  try {
+    dataset = static_cast<Dataset*>(datset_handle);
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:29]:
```suggestion
    dataset = static_cast<Dataset*>(dataset_handle);
```

File: tests/cpp_tests/test_serialize.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Serialize the reference
+    ByteBufferHandle buffer_handle;
+    int32_t buffer_len;
+    result = LGBM_DatasetSerializeReferenceToBinary(datset_handle, &buffer_handle, &buffer_len);
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:34]:
```suggestion
    result = LGBM_DatasetSerializeReferenceToBinary(dataset_handle, &buffer_handle, &buffer_len);
```

File: tests/cpp_tests/test_serialize.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    EXPECT_EQ(0, result) << "LGBM_DatasetSerializeReferenceToBinary result code: " << result;
+
+    ByteBuffer* buffer = nullptr;
+    Dataset* deserialized_dataset = nullptr;
+    try {
+      buffer = static_cast<ByteBuffer*>(buffer_handle);
+
+      // Deserialize the reference
+      DatasetHandle deserialized_datset_handle;
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:43]:
```suggestion
      DatasetHandle deserialized_dataset_handle;
```

File: tests/cpp_tests/test_serialize.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      result = LGBM_DatasetCreateFromSerializedReference(buffer->Data(),
+                                                         static_cast<int32_t>(buffer->GetSize()),
+                                                         dataset->num_data(),
+                                                         0,  // num_classes
+                                                         params,
+                                                         &deserialized_datset_handle);
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:49]:
```suggestion
                                                         &deserialized_dataset_handle);
```

File: tests/cpp_tests/test_serialize.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      EXPECT_EQ(0, result) << "LGBM_DatasetCreateFromSerializedReference result code: " << result;
+
+      // Confirm 1 successful API call
+      deserialized_dataset = static_cast<Dataset*>(deserialized_datset_handle);
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:53]:
```suggestion
      deserialized_dataset = static_cast<Dataset*>(deserialized_dataset_handle);
```

File: tests/cpp_tests/test_serialize.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+TEST(Serialization, JustWorks) {
+  // Load some test data
+  DatasetHandle datset_handle;
+  const char* params = "max_bin=15";
+  int result = TestUtils::LoadDatasetFromExamples("binary_classification/binary.train", params, &datset_handle);
+  EXPECT_EQ(0, result) << "LoadDatasetFromExamples result code: " << result;
+
+  Dataset* dataset;
+  try {
```
Comment (StrikerRUS) [inline tests/cpp_tests/test_serialize.cpp:28]: Doesn't `try/catch` here mean that test can silently fail and continue producing green CI job indicator even when everything is actually broken?

**Review by StrikerRUS — COMMENTED** [review]
Just some typos and one suggestion for the version constant.

File: src/io/dataset.cpp (hunk 2 above)
Comment (svotaw) [reply src/io/dataset.cpp:1085]: sure, this was just moved code, but I can improve it

File: src/io/dataset.cpp (hunk 1 above)
Comment (svotaw) [reply src/io/dataset.cpp:22]: Sure.  I made it fixed length, to simplify deserializing. (the example you gave is never deserialized)

File: tests/cpp_tests/test_serialize.cpp (hunk 10 above)
Comment (svotaw) [reply tests/cpp_tests/test_serialize.cpp:28]: ah, good catch, I had added that last minute to handle memory freeing.

Comment (StrikerRUS) [reply tests/cpp_tests/test_serialize.cpp:30]: OK, thanks for updating tests!
I guess we need the same in streaming tests:
https://github.com/microsoft/LightGBM/blob/c8712a9e083bfa0922f6e860f49a8a374816f43e/tests/cpp_tests/test_stream.cpp#L38
https://github.com/microsoft/LightGBM/blob/c8712a9e083bfa0922f6e860f49a8a374816f43e/tests/cpp_tests/test_stream.cpp#L145

**Review by StrikerRUS — COMMENTED** [review]
Thanks! I don't have any additional comments for this PR.

✔️

**Comment by guolinke** [comment]
@shiyu1994 can you help to reivew?

**Comment by svotaw** [comment]
@shiyu1994 can you take a look? ty

**Comment by jameslamb** [comment]
@shiyu1994 @guolinke can you help with a review on this?

**Comment by shiyu1994** [comment]
Sorry for the late response. Will review it within the next two days.

File: include/LightGBM/feature_group.h [hunk 11]
```diff
@@ -110,14 +110,56 @@ class FeatureGroup {
...
+    if (!local_used_indices.empty()) {
+      num_data = static_cast<data_size_t>(local_used_indices.size());
+    }
+    AllocateBins(num_data);
+
+    // Now load the actual data
+    if (is_multi_val_) {
+      for (int i = 0; i < num_feature_; ++i) {
+        multi_bin_data_.back()->LoadFromMemory(memory_ptr, local_used_indices);
+        memory_ptr += multi_bin_data_.back()->SizesInByte();
```
Comment (shiyu1994) [inline include/LightGBM/feature_group.h:137]:
```suggestion
        multi_bin_data_[i]->LoadFromMemory(memory_ptr, local_used_indices);
        memory_ptr += multi_bin_data_[i]->SizesInByte();
```

Comment (shiyu1994) [reply include/LightGBM/feature_group.h:137]: Otherwise, the data will always be loaded into the last bin.

File: src/io/dataset.cpp [hunk 12]
```diff
@@ -1099,6 +1033,117 @@ void Dataset::SaveBinaryFile(const char* bin_filename) {
   }
 }
 
+void Dataset::SerializeReference(ByteBuffer* buffer) {
+  Log::Info("Saving data to binary buffer");
```
Comment (shiyu1994) [inline src/io/dataset.cpp:1037]:
```suggestion
  Log::Info("Saving data reference to binary buffer");
```

**Review by shiyu1994 — COMMENTED** [review]
@svotaw Just left a few suggestions, could you please check? Thanks. Overall the PR LGTM.

File: include/LightGBM/feature_group.h (hunk 11 above)
Comment (svotaw) [reply include/LightGBM/feature_group.h:137]: nice! great catch

**Comment by svotaw** [comment]
@shiyu1994 I made the requested changes. Can you look it over and try rerunning the failures? they don't seem related to this PR

**Review by shiyu1994 — APPROVED** [review]

**Pull request merged**